fondant/component.py

```diff
@@ -1,260 +1,41 @@
-        return component_arguments
-
-    @abstractmethod
-    def _load_or_create_manifest(self) -> Manifest:
-        """Abstract method that returns the dataset manifest."""
-
-    @abstractmethod
-    def _process_dataset(self, manifest: Manifest) -> t.Union[None, dd.DataFrame]:
-        """Abstract method that processes the manifest and
-        returns another dataframe.
-        """
-
-    def _write_data(self, dataframe: dd.DataFrame, *, manifest: Manifest):
-        """Create a data writer given a manifest and writes out the index and subsets."""
-        data_writer = DaskDataWriter(manifest=manifest, component_spec=self.spec)
-        data_writer.write_dataframe(dataframe)
-
-    def run(self):
-        """Runs the component."""
-        input_manifest = self._load_or_create_manifest()
-
-        output_df = self._process_dataset(manifest=input_manifest)
-
-        output_manifest = input_manifest.evolve(component_spec=self.spec)
-
-        self._write_data(dataframe=output_df, manifest=output_manifest)
-
-        self.upload_manifest(output_manifest, save_path=self.output_manifest_path)
-
-    def upload_manifest(self, manifest: Manifest, save_path: str):
-        Path(save_path).parent.mkdir(parents=True, exist_ok=True)
-        manifest.to_file(save_path)
-
-
-class LoadComponent(Component):
-    """Base class for a Fondant load component."""
-
-    @staticmethod
-    def optional_fondant_arguments() -> t.List[str]:
-        return ["input_manifest_path"]
-
-    def _load_or_create_manifest(self) -> Manifest:
-        component_id = self.spec.name.lower().replace(" ", "_")
-        return Manifest.create(
-            base_path=self.metadata["base_path"],
-            run_id=self.metadata["run_id"],
-            component_id=component_id,
-        )
-
-    @abstractmethod
-    def load(self, *args, **kwargs) -> dd.DataFrame:
-        """Abstract method that loads the initial dataframe."""
-
-    def _process_dataset(self, manifest: Manifest) -> dd.DataFrame:
-        """This function loads the initial dataframe sing the user-provided `load` method.
-
-        Returns:
-            A `dd.DataFrame` instance with initial data'.
-        """
-        # Load the dataframe according to the custom function provided to the user
-        return self.load(**self.user_arguments)
-
-
-class TransformComponent(Component):
-    """Base class for a Fondant transform component."""
-
-    def _load_or_create_manifest(self) -> Manifest:
-        return Manifest.from_file(self.input_manifest_path)
-
-    @abstractmethod
-    def transform(self, *args, **kwargs) -> dd.DataFrame:
-        """
-        Abstract method for applying data transformations to the input dataframe.
-
-        Args:
-            args: The dataframe will be passed in as a positional argument
-            kwargs: Arguments provided to the component are passed as keyword arguments
-        """
+class BaseComponent:
+    def __init__(self, spec: ComponentSpec, **kwargs):
```

Comment:
I thought about that as well, but I'm not sure if only load and write components can be generic. I could see some transform components which could be generic as well. Eg. a `normalize_text` component, which normalizes every text column it is passed.

Reply:
Whether a component is generic or not is defined mainly in the `_write_data()` [method](https://github.com/ml6team/fondant/blob/00110d236eca63e3d01c561a5d225de2dbde8dee/fondant/executor.py#L155C3-L155C3), all components can be genetic since we're passing the spec to the executor with the `from_args()` method.  

I was mainly referring about whether to pass the spec to the `component` [link](https://github.com/ml6team/fondant/blob/00110d236eca63e3d01c561a5d225de2dbde8dee/fondant/executor.py#L162). This is currently only useful in the `write_to_hub` component since we have a custom logic on how to sink the data.

Reply:
Honestly, I am not sure if the current reusable components are generic. A lot of transform components containing hard coded column names. What actually means that I have to adapt my dataframe columns to match the component spec. Which requires for a user to know details about expected dataframe schema. 

From a usage perspective it would be nice to use a reusable component without worry about the required dataframe schema. 
E.g. I know there is a text normalization component. Now I want to pass my dataframe with text columns to the component. I would expected that I can (somehow) define on which columns the transformation will be applied. 
If I understood it correctly that is the idea behind: https://github.com/ml6team/fondant/issues/180

But isn't it a different topic?

Reply:
Yes actually we are working more and more with a "taxonomy" (predefined column names that each component should respect) in order to make them reusable. So the first component in a pipeline (typically the "load_from_hf_hub" right now) needs to map the existing columns to the format that we except. I do that [here](https://github.com/ml6team/fondant/blob/7f271246115568b487a6050f193003a38ddfe83f/examples/pipelines/datacomp/pipeline.py#L25-L35) for the Datacomp pipeline.